### ISSUE TITLE
openssh 6.9+: ssh-agent is never used by ssh

### ISSUE THREAD

**Issue by jbergstroem** [Homebrew/homebrew-dupes#488]
While using ssh on OS X, it never seems communicate with the bundled `ssh-agent` while using ssh.

(background) I removed the os x keychain patch a while ago for reasons similar to why it was just removed. I currently utilise `brew install keychain` for convenience but could as well just call `ssh-agent`. I've used `ssh-agent` for years in Linux, has never been a problem.

Lets add a few keys after verifying that we have the right environment (note, minor edits here and there)
```bash
$ set | grep -a SSH_AUTH && ps aux | grep ssh-agent
SSH_AUTH_SOCK=/var/folders/2d/qs_rvdcd4p38kvwgyql496tc0000gn/T//ssh-Np5NLkw8D59n/agent.1266
jbergstroem      1267   0.0  0.0  2462668    964   ??  Ss    1:01pm   0:00.02 ssh-agent
$ ssh-add .ssh/foo.key
Enter passphrase for key '~/.ssh/foo.key': 
$ ssh-add -l
2048 SHA256:asdf  ~/.ssh/foo.key (RSA)
$ grep -A 3 bar .ssh/config
Host bar
  Port 12345
  ForwardAgent yes
  IdentityFile ~/.ssh/foo.key
$ ssh bar
Enter passphrase for key '~/.ssh/foo.key': 
```

I've used dtrace, dtruss, etc to try to figure out what's going on and the only conclusion is that ssh just ignores ssh-agent. Entering my password for the key works just fine.

Also, using ForwardAgent and `ssh-agent` from the host I just ssh:ed to works "Just Fine". Am I missing something about how ssh is built on OS X? The brewfile leaves me pretty short.

**Comment by croessner** [comment]
I have a similar problem. After rebooting the Mac, I can not login to my servers. I checked the environment for SSH_AGENT, but even there was a listed socket, there was no process running. I had to start ssh-agent by hand and had to add my keys.

**Comment by jbergstroem** [comment]
@croessner sounds like a different issue though; I'm not after launching `ssh-agent` at launch; you can just create a launchd script to do that.

**Comment by croessner** [comment]
Yes, you are right. I found out that keychain support was dropped. So I went back to Apples ssh

**Comment by jbergstroem** [comment]
for the record, you're not tied to apple's ssh-agent just because you want a convenient init script; just copy/edit theirs.

**Comment by croessner** [comment]
The patch has been removed. So editing /System/Library/...plist does not work anymore for me. Of course I could use ssh-agent with eval in .bash_profile and enter passphrase once at each boot, but this is not an option for me.

**Comment by jbergstroem** [comment]
@croessner we're moving off-topic here, but the patch isn't relevant for the init script - rather for having ssh-agent integrating with keychain. If you want to keep talking, lets do it somewhere else. Feel free to ping me at irc: jbergstroem@freenode - otherwise there's probably a few guides for how to create your own launchd init script on google.

**Comment by timbru31** [comment]
I'm having the same problem. After the recent upgrade yesterday ```ssh-add ~/.ssh/id_rsa``` was not working anymore. It turns out that the ssh-agent is no longer started on boot or when a new shell instance is launched?
eval `ssh-agent -s` seems to fix the problem.

**Comment by grahame** [comment]
I think the problem is that 'ssh-agent -l' is no longer supported; launchd is trying to run it with that argument.

**Comment by bd808** [comment]
Related: https://github.com/Homebrew/homebrew-dupes/pull/482#issuecomment-118994372

**Comment by jbergstroem** [comment]
Just to make it super clear: My problem is not that ssh-agent isn't launching, it's that ssh doesn't communicate with it forcing me to enter my password.

**Comment by DomT4** [comment]
Have been looking into this, and comparing it to the way Debian build. There's not a huge difference.

The behaviour is similar to things that store a built-in `$PATH` at compile time and then glue itself to that instead of to the users runtime `$PATH` - This is important in Homebrew as we modify the `$PATH` before compile.

`sshd_config` does store the compile `$PATH`, but I'm not sure on the relevance of this yet.

There's an awful lot of documentation on the agent, but the assumption behind a lot of documentation is that there won't be a competing agent built into the system (Or you'll be replacing that), and obviously on OS X there is and we don't overwrite the system installation.

**Comment by jbergstroem** [comment]
@DomT4 just to clarify; I've unloaded the system plist and only start my `ssh-agent` manually through `keychain`. You mention a few interesting things in terms of how ssh is compiled, I had a quick look through ssh and ssh-agent and it seems they both use the same function to check for the socket. What I don't find is the references to the compile path. How would this affect above scenario? Also, thanks for taking the time to look into it.

**Comment by tduzan** [comment]
I believe this issue may be related to a discovery I made when working up a patch to the 'ssh' module for prezto as a workaround for the removal of the patch in Homebrew.  By default, your profile loads some stuff from the system to the environment before loading your own user profile environment.  Part of this is some things which get set by launchd/launchctl.  One of those things is the SSH_AUTH_SOCK environment variable, which defaults to pointing to a socket controlled by launchd which gets redirected to the system default ssh-agent.  If you are using brewed OpenSSH, when you run 'ssh' you're running OpenSSH from Homebrew which doesn't have the customizations to support the behavior that launchd expects, so it fails to communicate with the system ssh-agent and moves on.  You need to ensure that the SSH_AUTH_SOCK gets set to a stored persistent socket on each session you load (e.g. new tab in Terminal.app or re-opening Terminal.app).

You can check the implementation in my fork of prezto at https://github.com/Tristor/prezto/blob/master/modules/ssh/init.zsh which does indeed work.  On an initial shell after reboot, ssh-agent will prompt me for my key passphrases, after which I have a persistent agent and socket running in the background that get used by every shell session and the 'ssh' command works as expected utilizing loaded identities referenced from my .zshrc.  If you're not using ZSH + prezto, then this change can't be used as is, but the scripting should be understandable enough to replicate for your particular configuration.

**Comment by jbergstroem** [comment]
@tduzan I use `keychain`  (brew install keychain) to set my environment for every session.

Anyway, just switched to el capitan which seems to have a recent enough version of ssh bundled (OpenSSH_6.7p1, LibreSSL 2.1.3). Problem "solved".

**Comment by leonklingele** [comment]
@jbergstroem Slightly off-topic:
So El Capitan's OpenSSH in compiled against LibreSSL? That's actually quite nice.
Can you do a quick `openssl version` for me, please? :)

**Comment by jbergstroem** [comment]
@leonklingele:
``` bash
~ $ openssl version
OpenSSL 0.9.8zf 19 Mar 2015
```

**Comment by mistydemeo** [comment]
10.11 installs a private copy of the libressl libraries in `/usr/lib/ssh`, without the openssl binary or any headers. It's possible this will change in later OS X versions but they seem to be taking it slow.

**Comment by jbergstroem** [comment]
we are moving off topic here, so this'll be my last post about libre -- but I don't think Apple aims to replace openssl. They are two different libraries and [not meant to be drop-in replacements](https://blog.flameeyes.eu/2015/06/libressl-openssl-collisions-and-problems#gsc.tab=0).

**Comment by Sayhar** [comment]
Hello, I'm still having this issue. Was it ever fixed?

**Comment by yveslaroche** [comment]
I'm on OSX 11.11 and using the systems SSH-agent with the Homebrew OpenSSH and as a result the passphrases are still stored in the Keychain. OpenSSH is installed with LibreSSL.
`❯ brew install openssh --with-libressl`

```sh
❯ ps x | grep ssh-agent
57824   ??  S      0:00.18 /usr/bin/ssh-agent -l
58238 s000  S+     0:00.00 grep --color=auto ssh-agent

~
❯ which ssh
/usr/local/bin/ssh

~
❯ ssh -V
OpenSSH_7.1p2, LibreSSL 2.3.2
```

**Comment by yveslaroche** [comment]
Forgot to mention, in order to add your passphrase to the Keychain you have to use the OSX OpenSSH once per key.
```
❯ /usr/bin/ssh -i ~/.ssh/your_private_key your_server
```

**Comment by aeuae** [comment]
> Forgot to mention, in order to add your passphrase to the Keychain you have to use the OSX OpenSSH once per key.

 @yveslaroche

 thank you, works flawlessly. 

This workaround is what I have been looking for!

**Comment by joliver** [comment]
Maybe I'm coming way late to the game, but I'm able to use the OS X keychain and the homebrew compiled SSH without a problem. Here's my environment:

OS X 10.11.3 (El Capitan)

```
$ /usr/bin/ssh -V
OpenSSH_6.9p1, LibreSSL 2.1.8
```

```
$ /usr/local/bin -V
OpenSSH_7.1p2, OpenSSL 1.0.2f  28 Jan 2016
```

```
$ brew --version
Homebrew 0.9.5 (git revision 5ae4; last commit 2016-02-18)
```

At installation of homebrew, I ran `brew tap homebrew/dupes` which is currently at git commit `7a70972f8dc`

To install openssh, I ran `brew install openssh` without any special flags. I don't have anything else installed. This is a fresh system.

The `ssh-agent` application isn't yet running according to `ps aux | grep ssh-agent`

I then copied my encrypted ssh `id_rsa` to my ~/.ssh directory (which I had to create on the fresh install).

For the following steps, be sure to run `which ssh` to know if you're running the built-in SSH or the homebrew installed version.

Notice that for this invocation of the homebrew version of SSH, it will currently prompt for a password.
```
$ /usr/local/bin/ssh me@server
Enter passphrase for key '/Users/me/.ssh/id_rsa':
```

Here's where the magic happens:
```
$ /usr/bin/ssh me@server.com
```
At this point, I should get the built-in OS X dialog popup window outside of the terminal asking me for my passphrase. I provide my passphrase and check the "Remember password in my keychain" option.

From there, I can login using the homebrew SSH version and it will not prompt me for a password.
```
$ /usr/local/bin/ssh me@server
```

Alternatively, I can run `$ /usr/bin/ssh-add -K /Users/me/.ssh/id_rsa` which will prompt for my password in the terminal and then store the passphrase in the OS X Keychain Access program.

Either method will survive reboots so that I don't have to type my password in (other than my login to OS X password) whenever I SSH into a remote machine.

Running `ps aux | grep ssh-agent` after SSH-ing the first time shows that builtin OS X agent located at `/usr/bin/ssh-agent` is running.

My key is an RSA key, not an elliptic curve private key.
